<a href="https://colab.research.google.com/github/davidofitaly/machine-learning-notes/blob/main/supervised/02_regression/01_linear_regression_normal_equation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###import of libraries

In [1]:
import numpy as np

###data preparation

In [2]:
X1 = np.array([1,2,3,4,5,6])
Y = np.array([3000,3250,3500,3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [3]:
X1 = X1.reshape(m, 1)
print(X1)
print(X1.shape)

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]
(6, 1)


In [4]:
bias = np.ones((m, 1))
print(bias)
print(bias.shape)


[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]
(6, 1)


In [5]:
X = np.append(bias, X1, axis=1)
print(X)
print(X.shape)

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]
(6, 2)


### <a name='2'></a> Równanie normalne

Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$W = (X^TX)^{-1}(X^TY)$$

In [6]:
np.dot(X.T, X)

array([[ 6., 21.],
       [21., 91.]])

In [7]:
L = np.linalg.inv(np.dot(X.T,X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [8]:
P = np.dot(X.T,Y)
P

array([21750., 80500.])

In [9]:
np.dot(L,P)

array([2750.,  250.])

The final form of the model$$Y = 2750 + 250X_1$$

###Linear regression using scikit-learn

In [12]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1,Y)

print(regression.intercept_)
print(regression.coef_[0])

2750.0
250.0
